examples/advanced/visual_sudoku.py

```diff
@@ -0,0 +1,281 @@
+
+    # Dictionary (number, corresponding sample images)
+    digit_indices = {number: torch.LongTensor(*np.where(testset.targets == number)) for number in range(1,10)}
+
+    # replace all numeric values in sudoku torch tensor grid by an mnist digit
+    
+    # TODO: Why do you not use np.ndenumerate ? and then for each number in the 
+    #       grid you loop select a random image?
+    # IS it for having differnet images ? if yes, motivate or comment :)
+    # TODO: PROPOSED:
+    for idx, val in np.ndenumerate(sudoku_puzzle[sudoku_puzzle > 0]):
+        image_tensor = np.random.choice(digit_indices[val])
+        vizsudoku[idx] = image_tensor
+
+    ## OLD:
+    for val in np.unique(sudoku_puzzle[sudoku_puzzle > 0]):
+
+        val_idx = np.where(sudoku_puzzle == val)
+        idx = torch.LongTensor(np.random.choice(digit_indices[val], len(sudoku_puzzle[val_idx])))
+        vizsudoku[val_idx] = torch.stack([testset[i][0] for i in idx])
+
+    return vizsudoku
+
+
+
+from torch import nn 
+import torch.nn.functional as F
+
+class LeNet(nn.Module):
+    '''
+        TODO: Write doc+ reference to paper
+    '''
+    def __init__(self):
+        super(LeNet, self).__init__()
+        self.conv1 = nn.Conv2d(1, 6, 5, 1, padding=2)
+        self.conv2 = nn.Conv2d(6, 16, 5, 1)
+        self.fc1 = nn.Linear(5*5*16, 120)
+        self.fc2 = nn.Linear(120, 84)
+        self.fc3 = nn.Linear(84,10)
+
+    def forward(self, x):
+        x = F.relu(self.conv1(x))
+        x = F.max_pool2d(x, 2, 2)
+        x = F.relu(self.conv2(x))
+        x = F.max_pool2d(x, 2, 2)
+        x = x.view(-1, 5*5*16) 
+        x = F.relu(self.fc1(x))
+        x = self.fc2(x)
+        x = self.fc3(x)
+        return F.log_softmax(x, dim=1)
+
+def load_clf(clf_classname, path):
+    '''
+        TODO: Write doc.
+    '''
+    net = clf_classname()
+    state_dict = torch.load(path, map_location=lambda storage, loc: storage)
+    net.load_state_dict(state_dict)
+    return net
+
+
+@torch.no_grad() #TODO: What does this do ?
+def predict_proba_sudoku(model, vizsudoku):
+    #TODO: Change shape of output to work for differnet kinds of sudoku shapes (4x4, 9x9, ...)
+    # reshape from 9x9x1x28x28 to 81x1x28x28
+    pred = model(vizsudoku.flatten(0,1))
+    # our NN returns 81 probabilistic vector: an 81x10 matrix
+    return pred.reshape(9,9,10).detach() # reshape as 9x9x10 tensor for easier visualisation
+
+if __name__ == '__main__':
+    '''
+        Missing a bit of explanations even small bit
+    '''
+    e = 0 # empty cell
+    sample_sudoku_grid = np.array([
+        [e, e, e,  2, e, 5,  e, e, e],
+        [e, 9, e,  e, e, e,  7, 3, e],
+        [e, e, 2,  e, e, 9,  e, 6, e],
+
+        [2, e, e,  e, e, e,  4, e, 9],
+        [e, e, e,  e, 7, e,  e, e, e],
+        [6, e, 9,  e, e, e,  e, e, 1],
+
+        [e, 8, e,  4, e, e,  1, e, e],
+        [e, 6, 3,  e, e, e,  e, 8, e],
+        [e, e, e,  6, e, 8,  e, e, e]
+    ])
+
+    # Emilio: Missing a bit of context here ?
```

Comment:
idd, also watch your variable names, you call it 'sample_sudoku_grid' which you then feed into a 'sample_...' function which means somethign else.

Also higher up, avoid 'model' for the NN as 'model' in CP/CPMpy is smth different...